gate-web/src/main/groovy/com/netflix/spinnaker/gate/security/rolesprovider/github/GithubTeamsUserRolesProvider.groovy

```diff
@@ -0,0 +1,130 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.gate.security.rolesprovider.github
+
+import com.netflix.spinnaker.gate.config.GitHubProperties
+import com.netflix.spinnaker.gate.security.rolesprovider.UserRolesProvider
+import com.netflix.spinnaker.gate.security.rolesprovider.github.client.GitHubMaster
+
+import com.netflix.spinnaker.gate.security.rolesprovider.github.client.model.Team
+import com.netflix.spinnaker.gate.security.rolesprovider.github.client.model.TeamMembership
+import groovy.util.logging.Slf4j
+import org.springframework.beans.factory.InitializingBean
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
+import org.springframework.stereotype.Component
+import org.springframework.util.Assert
+import retrofit.RetrofitError
+import retrofit.client.Response
+
+@Slf4j
+@Component
+@ConditionalOnProperty(value = "auth.groupMembership.service", havingValue = "github")
+class GithubTeamsUserRolesProvider implements UserRolesProvider, InitializingBean {
+
+  @Autowired
+  GitHubMaster master
+
+  @Autowired
+  GitHubProperties gitHubProperties
+
+  @Override
+  Collection<String> loadRoles(String userName) {
+    if (!userName) {
+      return []
+    }
+    // check organization if set.
+    // If organization is unset, all GitHub users can login and have full access
+    // If an organization is set, add it to roles to restrict users to this organization
+    // If organization is set AND requiredGroupMembership set, organization members will have RO access
+    // and requiredGroupMembership members RW access
+    if (!gitHubProperties.organization) {
+      return []
+    } else {
+      Boolean isMemberOfOrg = false
+
+      try {
+        Response response = master.gitHubClient.isMemberOfOrganization(gitHubProperties.organization, userName)
+        isMemberOfOrg = (response.status == 204)
+      } catch (RetrofitError e) {
+        if (e.getKind() == RetrofitError.Kind.NETWORK) {
+          log.error("Could not find the server ${master.baseUrl}", e)
+          return []
+        } else if (e.response.status == 404) {
+          return []
+        }
+      }
+      if (isMemberOfOrg) {
+        List result = []
+        result.add(gitHubProperties.organization)
+
+        // Get teams of the current user
+        List<Team> teams
+        try {
+          teams = master.gitHubClient.getOrgTeams(gitHubProperties.organization)
+        } catch (RetrofitError e) {
+          if (e.getKind() == RetrofitError.Kind.NETWORK) {
+            log.error("Could not find the server ${master.baseUrl}", e)
+          } else if (e.response.status == 404) {
+            return result
+          }
+        }
+        if (teams) {
+          teams.each { Team t -> if (isMemberOfTeam(t, userName)) result.add(t.slug) }
```

Comment:
You might consider a little more condense version:
```
result = teams.findAll { isMemberOfTeam(it, username) }
// then add L73 afterwards)
result.add(gitHubProperties.organization)
```
But i'm not sure if the ordering matters?